spec/lib/granite/action_spec.rb

```diff
@@ -169,4 +169,45 @@ def execute_perform!(*)
       end
     end
   end
+
+  describe '#merge_errors' do
+    subject(:action) { Action.new }
+
+    let(:errors_to_merge) { ActiveModel::Errors.new(Action.new) }
+
+    before { action.errors.add(:base, 'some error') }
+
+    context 'when errors have different attributes' do
+      before { errors_to_merge.add(:name, 'some error') }
+
+      it 'merges given errors with existing ones' do
+        action.merge_errors(errors_to_merge)
+
+        expect(action.errors.count).to eq(2)
+        expect(action.errors.messages).to match(base: ['some error'], name: ['some error'])
+      end
+    end
+
+    context 'when errors have different types' do
+      before { errors_to_merge.add(:base, 'some other error') }
+
+      it 'merges given errors with existing ones' do
+        action.merge_errors(errors_to_merge)
+
+        expect(action.errors.count).to eq(2)
+        expect(action.errors.messages).to match(base: ['some error', 'some other error'])
+      end
+    end
+
+    context 'when errors are duplicates' do
+      before { errors_to_merge.add(:base, 'some error') }
+
+      it 'does not duplicate existing errors' do
+        action.merge_errors(errors_to_merge)
+
+        expect(action.errors.count).to eq(1)
+        expect(action.errors.messages).to match(base: ['some error'])
+      end
+    end
```

Comment:
I think you're missing cases where `error.options` are different and where `error.options` are same?

Reply:
@andriusch TBH I'm not even sure if we need to check `added?` with options, or `attribute` and `type` is enough.

Reply:
Are `options` what is passed to translations? If yes, then I think we need to check them. E.g. `errors.add(:base, :invalid, name: 'John')` generates different message from `errors.add(:base, :invalid, name: 'Peter')`. Though I hardly imagine a realistic case where it would matter.

Reply:
Yes, option even can contain `message`. And yes in real-life use-cases it doesn't matter, at least from what I see now. So alright, let's keep them and I'll add also specs for them.